src/aspire/orientation/__init__.py

```diff
@@ -7,35 +15,434 @@ class Orient3D:
+        idx_i = np.array(idx_i, dtype='int')
+        idx_j = np.array(idx_j, dtype='int')
+        rp = np.random.choice(np.arange(len(idx_j)), size=n_equations, replace=False)
+
+        # might be able to vectorize this
+        for shift_eq_idx in range(n_equations):
+            i = idx_i[rp[shift_eq_idx]]
+            j = idx_j[rp[shift_eq_idx]]
+
+            r_i = rotations[:, :, i]
+            r_j = rotations[:, :, j]
+            c_ij, c_ji = self._common_line_r(r_i.T, r_j.T, 2 * n_theta)
+
+            if c_ij >= n_theta:
+                c_ij -= n_theta
+                c_ji -= n_theta
+            if c_ji < 0:
+                c_ji += 2 * n_theta
+
+            c_ij = int(c_ij)
+            c_ji = int(c_ji)
+            is_pf_j_flipped = 0
+            if c_ji < n_theta:
+                pf_j = pf[:, c_ji, j]
+            else:
+                pf_j = pf[:, c_ji - n_theta, j]
+                is_pf_j_flipped = 1
+            pf_i = pf[:, c_ij, i]
+
+            pf_i *= h
+            pf_i[r_max - 1:r_max + 2] = 0
+            pf_i /= np.linalg.norm(pf_i)
+            pf_i = pf_i[:r_max]
+
+            pf_j *= h
+            pf_j[r_max - 1:r_max + 2] = 0
+            pf_j /= np.linalg.norm(pf_j)
+            pf_j = pf_j[:r_max]
+
+            pf_i_flipped = np.conj(pf_i)
+            pf_i_stack = np.einsum('i, ij -> ij', pf_i, shift_phases)
+            pf_i_flipped_stack = np.einsum('i, ij -> ij', pf_i_flipped, shift_phases)
+
+            c1 = 2 * np.real(np.dot(np.conj(pf_i_stack.T), pf_j))
+            c2 = 2 * np.real(np.dot(np.conj(pf_i_flipped_stack.T), pf_j))
+
+            sidx1 = np.argmax(c1)
+            sidx2 = np.argmax(c2)
+
+            if c1[sidx1] > c2[sidx2]:
+                dx = -max_shift + sidx1 * shift_step
+            else:
+                dx = -max_shift + sidx2 * shift_step
+
+            idx = np.arange(4 * shift_eq_idx, 4 * shift_eq_idx + 4)
+            shift_alpha = c_ij * d_theta
+            shift_beta = c_ji * d_theta
+            shift_i[idx] = shift_eq_idx
+            shift_j[idx] = [2 * i, 2 * i + 1, 2 * j, 2 * j + 1]
+            shift_b[shift_eq_idx] = dx
+
+            # check with compare cl
+            if not is_pf_j_flipped:
+                shift_eq[idx] = [np.sin(shift_alpha), np.cos(shift_alpha),
+                                 -np.sin(shift_beta), -np.cos(shift_beta)]
+            else:
+                shift_beta -= np.pi
+                shift_eq[idx] = [-np.sin(shift_alpha), -np.cos(shift_alpha),
+                                 -np.sin(shift_beta), -np.cos(shift_beta)]
+
+        t = 4 * n_equations
+        shift_eq[t: t + n_equations] = shift_b
+        shift_i[t: t + n_equations] = np.arange(n_equations)
+        shift_j[t: t + n_equations] = 2 * n_projs
+        tmp = np.where(shift_eq != 0)[0]
+        shift_eq = shift_eq[tmp]
+        shift_i = shift_i[tmp]
+        shift_j = shift_j[tmp]
+        shift_equations = sps.csr_matrix((shift_eq, (shift_i, shift_j)),
```

Comment:
The `shift_equations` in `build_clmatrix` are built from the common matrix estimated from images but it looks that they are not used for estimate shifts in real case. I have separated them into a new function in commit b2d7dcc. The `shift_equations` in `estimate_shifts` are used in real case and calculated from the estimated rotations based on some approximation of memory factor. More improvements and comments have been added in commit  bde3c46.

Reply:
It is not used in the synchronization matrix method but I an not sure that it is not used for other methods in the original Matlab code. And also there are additional assumptions to reduce the number of equations in the `estimate_shifts`. I think it is better to keep it and add more comments to show the differences.

Reply:
No we should only keep one version. Check to see if the other version is used elsewhere in the Matlab version. If not, we remove the unused one. If it is, let's figure out if the differences are important enough to keep. It seems to me that there's a way to avoid reducing the number of equations in `estimate_shifts` by setting `memory_factor` large enough, which would give the same result as the calculations in `build_clmatrix`.

Reply:
They are different methods to get shift equations, one is from the common line matrix without rotations, the other is from the estimated rotations.  I have separated the function of `get_shift_equations` for `estimate_shifts` in commit b9a782f so that both methods can be called.

Reply:
More updates have been done on commits e78a4cb, 14333e2, and b9a012a.

Reply:
We are not keeping two versions of the same function, especially if one is never used. I don't understand why this is still an issue.

Reply:
The method generating shift equations from the common lines matrix has been removed in commit 7db94d1.